### PR TITLE
[CodeComplete] Mark results as not recommended if global actor context doesn't match

### PR DESCRIPTION
**Pull request by ahoppen**
Annotate cross-actor references to global actors as `async` and mark them as not-recommended if they must be executed on a different actor than the current context is on.

Resolves rdar://75902598

### CODE REVIEW

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please test

File: lib/Sema/TypeCheckStmt.cpp [hunk 1]
```diff
@@ -1949,6 +1949,7 @@ bool TypeCheckASTNodeAtLocRequest::evaluate(Evaluator &evaluator,
   if (auto CE = dyn_cast<ClosureExpr>(DC)) {
     if (CE->getBodyState() == ClosureExpr::BodyState::Parsed) {
       swift::typeCheckASTNodeAtLoc(CE->getParent(), CE->getLoc());
+      checkClosureActorIsolation(CE);
```
Comment (rintaro) [inline lib/Sema/TypeCheckStmt.cpp:1952]: I'm concerned that `ActorIsolationChecker` walk into unrelated sibling closures or the children of this closure.
How about exposing `ActorIsolationChecker::determineClosureIsolation()` and call it directly?
```cpp
ActorIsolationChecker checker(closure->getParent());
CE->setActorIsolation(checker->determineClosureIsolation(CE));
```

Comment (rintaro) [reply lib/Sema/TypeCheckStmt.cpp:1952]: Or, how about calling `checkFunctionActorIsolation()` (or `checkTopLevelActorIsolation()`) in `ide::typeCheckContextAt()` right after calling `typeCheckASTNodeAtLoc()`?

File: lib/IDE/Refactoring.cpp [hunk 2]
```diff
@@ -4571,7 +4571,8 @@ struct CallbackCondition {
   /// }
   /// ```
   CallbackCondition(const Decl *Subject, const CaseLabelItem *CaseItem) {
-    if (auto *EEP = dyn_cast<EnumElementPattern>(CaseItem->getPattern())) {
+    if (auto *EEP = dyn_cast<EnumElementPattern>(
+            CaseItem->getPattern()->getSemanticsProvidingPattern())) {
```
Comment (rintaro) [inline lib/IDE/Refactoring.cpp:4575]: Unrelated changes?

Comment (ahoppen) [reply lib/IDE/Refactoring.cpp:4575]: Oh, yes. I’ll remove it.

File: lib/Sema/TypeCheckStmt.cpp (hunk 1 above)
Comment (ahoppen) [reply lib/Sema/TypeCheckStmt.cpp:1952]: Thanks. I didn’t realize that `TypeCheckASTNodeAtLocRequest` was called multiple times. 

I think it’ll be good to just set the current closure’s actor isolation as you proposed in the first comment because it minimizes the amount of actor isolation checking that needs to be done. I’ll update it.

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please test macOS

**Comment by ahoppen** [comment]
@swift-ci Please test

File: test/IDE/complete_global_actorisolation.swift [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+    }
+  }
+
+  func testInLocalFunc() {
+    @MyGlobalActor func localFunc() {
+      let _ = otherInstanceOfMyClass.#^IN_LOCAL_FUNC_ON_GLOBAL_ACTOR_OTHER_DOT?check=IN_FUNC_ON_GLOBAL_ACTOR_SELF_DOT^#
+    }
+  }
+}
+
```
Comment (rintaro) [inline test/IDE/complete_global_actorisolation.swift:162]: Could you add a nested single expression closure test? e.g.
```swift
@MyGlobalActor func globalFuncOnMyGlobalActor() {}
@MyOtherGlobalActor func globalFuncOnMyOtherGlobalActor() {}

func foo<T>(_: () async -> T) {}

@MyGlobalActor func test() {
    foo {
        foo {
            <HERE>
        }
    }
}
```

File: test/IDE/complete_global_actorisolation.swift [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+class MyNonSendable {}
+struct MySendable {}
+
+actor MyActor {}
+
+@globalActor
+class MyGlobalActor {
+  static var shared = MyActor()
+}
```
[the remark this answers is not in the document]
Comment (ahoppen) [reply test/IDE/complete_global_actorisolation.swift:12]: I didn’t realize that. Thanks!

**Comment by ahoppen** [comment]
@swift-ci Please test

File: test/IDE/complete_global_actorisolation.swift [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+// RUN: %target-swift-ide-test -batch-code-completion -source-filename %s -filecheck %raw-FileCheck -completion-output-dir %t -enable-experimental-concurrency
+
+class MyNonSendable {}
+struct MySendable {}
+
+actor MyActor {}
+
+@globalActor
+actor MyGlobalActor {
+  static var shared = MyActor()
```
Comment (rintaro) [inline test/IDE/complete_global_actorisolation.swift:11]: I believe this must be `MyGlobalActor` in the new syntax.

> The type of `shared` must be of the enclosing actor type

Comment (ahoppen) [reply test/IDE/complete_global_actorisolation.swift]: Sorry, I missed that part. Fixed it now.

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please test macOS

**Review by rintaro — APPROVED** [review]

**Comment by ahoppen** [comment]
@swift-ci Please test macOS

**Comment by ahoppen** [comment]
@swift-ci Please smoke test macOS

File: lib/Sema/TypeCheckStmt.cpp [hunk 6]
```diff
@@ -1949,6 +1949,19 @@ bool TypeCheckASTNodeAtLocRequest::evaluate(Evaluator &evaluator,
...
+      // from outermost to innermost, we don't want to call checkActorIsolation,
+      // because that would cause actor isolation to be checked multiple times
+      // for nested closures. Instead, call determineClosureActorIsolation
+      // directly and set the closure's actor isolation manually. We can
+      // guarantee of that the actor isolation of enclosing closures have their
+      // isolation checked before nested ones are being checked by the way
+      // TypeCheckASTNodeAtLocRequest is called multiple times, as described
+      // above.
+      auto ActorIsolation = determineClosureActorIsolation(CE);
+      CE->setActorIsolation(ActorIsolation);
```
Comment (DougGregor) [inline lib/Sema/TypeCheckStmt.cpp:1964]: Hmm, it feels a bit dicey setting the actor isolation for a closure in more than one place. Should we request'ify `determineClosureActorIsolation`?

Comment (ahoppen) [reply lib/Sema/TypeCheckStmt.cpp:1964]: I agree that it’s not really pretty. But the problem is that the `ActorIsolationChecker` currently implicitly assumes that it’s walking the tree top to bottom. 

IMHO the proper solution would be to completely requesting `ActorIsolationChecker` so that it works bottom to top, requesting the actor isolation of an outer type as needed. But I felt that was too big of a change for Swift 5.5 and that’s why I stuck with the current, more local, approach.
